File: backend-golang/modules/payments/internal/webhook.go
// Real Stripe webhook signature verification (HMAC-SHA256 + timestamp
// tolerance), following the same algorithm as backend-python's
// webhook_handler.py — see docs/stripe-webhooks.md for the shared pattern.
package internal

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

const defaultToleranceSeconds = 300

func VerifyStripeWebhook(payload []byte, sigHeader string) error {
	secret := os.Getenv("STRIPE_WEBHOOK_SECRET")
	if secret == "" {
		return errors.New("STRIPE_WEBHOOK_SECRET is not configured")
	}

	var timestamp int64
	var sigs []string
	for _, part := range strings.Split(sigHeader, ",") {
		kv := strings.SplitN(part, "=", 2)
		if len(kv) != 2 {
			continue
		}
		switch strings.TrimSpace(kv[0]) {
		case "t":
			timestamp, _ = strconv.ParseInt(kv[1], 10, 64)
		case "v1":
			sigs = append(sigs, kv[1])
		}
	}
	if timestamp == 0 || len(sigs) == 0 {
		return errors.New("malformed Stripe-Signature header")
	}

	now := time.Now().Unix()
	if diff := now - timestamp; diff > defaultToleranceSeconds || diff < -defaultToleranceSeconds {
		return errors.New("timestamp outside tolerance — possible replay")
	}

	signedPayload := fmt.Sprintf("%d.%s", timestamp, payload)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(signedPayload))
	expected := hex.EncodeToString(mac.Sum(nil))

	for _, sig := range sigs {
		if hmac.Equal([]byte(expected), []byte(sig)) {
			return nil
		}
	}
	return errors.New("signature mismatch")
}

File: backend-rust/src/modules/payments/internal/webhook.rs
//! Real Stripe webhook signature verification (HMAC-SHA256 + timestamp
//! tolerance) — mirrors backend-python's webhook_handler.py. See
//! docs/stripe-webhooks.md for the shared cross-language pattern.
use hmac::{Hmac, Mac};
use sha2::Sha256;
use std::time::{SystemTime, UNIX_EPOCH};

const TOLERANCE_SECS: i64 = 300;

pub fn verify_stripe_webhook(payload: &[u8], sig_header: &str, secret: &str) -> Result<(), String> {
    if secret.is_empty() {
        return Err("STRIPE_WEBHOOK_SECRET is not configured".into());
    }
    let mut timestamp: i64 = 0;
    let mut sigs = Vec::new();
    for part in sig_header.split(',') {
        if let Some((k, v)) = part.split_once('=') {
            match k.trim() {
                "t" => timestamp = v.parse().unwrap_or(0),
                "v1" => sigs.push(v.to_string()),
                _ => {}
            }
        }
    }
    if timestamp == 0 || sigs.is_empty() {
        return Err("malformed Stripe-Signature header".into());
    }
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
    if (now - timestamp).abs() > TOLERANCE_SECS {
        return Err("timestamp outside tolerance — possible replay".into());
    }
    let signed_payload = format!("{}.{}", timestamp, String::from_utf8_lossy(payload));
    let mut mac = Hmac::<Sha256>::new_from_slice(secret.as_bytes()).unwrap();
    mac.update(signed_payload.as_bytes());
    let expected = hex::encode(mac.finalize().into_bytes());
    if sigs.iter().any(|s| s == &expected) {
        Ok(())
    } else {
        Err("signature mismatch".into())
    }
}
